### PR TITLE
Simplify matching decoding

### LINKED ISSUE #413: Simplify matching decoding
### What is the expected enhancement?

Do matching with a simple, code agnostic wrapper around pymatching.

Also modify the decoding graph to support this, including distinguishing between boundary and logical nodes

### PR DESCRIPTION
**Pull request by quantumjim**
### Summary

fixes #413

Removes legacy matching decoders, and replaces them with a thin layer over PyMatching.

### Details and comments

* Separates logical nodes and boundary nodes into two separate concepts.
* Allows for two ways for the decoding graph to determine the effect on logicals: logical nodes and graphs with a fault id.
* Introduces a lightweight matching decoder: a wrapper around PyMatching.
* Removes old matching decoders.

### CODE REVIEW

File: extern/pybind11/tests/test_gil_scoped.py
Comment (grace-harper) [inline extern/pybind11/tests/test_gil_scoped.py:1]: Hey, just to check, did you make code modifications to `pybind11` or just the reformatting? 
If it's just the reformatting, I think we should split this PR into 2, one for all the Simplify Matching Decoding stuff and one for the reformatting of `pybind11` 
(Fwiw, given that `pybind11` is in `extern` and perhaps should even be a submodule, I don't personally feel like we necessarily have to reformat it, but definitely let me know if you feel differently)

**Review by grace-harper — CHANGES_REQUESTED** [review]
Just wanted to check in about the pybind11. Everything else looks good. The pymatching integration is super cool.

File: extern/pybind11/tests/test_gil_scoped.py
Comment (quantumjim) [reply extern/pybind11/tests/test_gil_scoped.py:1]: The reformatting of that folder was unintentional, so I reverted it.

**Review by grace-harper — APPROVED** [review]
LGTM

**Comment by quantumjim** [comment]
Since #420 was based on this branch, that PR also covers all the changes here

**Pull request closed without merging**